### PR TITLE
raft: stop resetting electionElapsed when attempting to campaign

### PR DESCRIPTION
**Pull request by iskettaneh**
This commit changes the current behaviour where we reset electionElapsed when campaigning (to allow us to campaign again on the next randomizedElectionTimeout). This problem with the current behaviour is that if a follower can't campaign (due to store liveness limitations), resetting the electionElapsed will re-arm the inHeartbeatLease, making the follower reject MsgVotes.

This commit changes that by campaigning at every randomizedElectionTimeout interval.

Epic: None

Release note: None

### CODE REVIEW

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OA04JmO82d670ymR_Gb:-OA04JmO82d670ymR_Gc:b6ktb9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/353fa4043b1431eb8268604a898f4ccef398d5a9/pkg/raft/raft.go#L1128)):*
> ```Go
> 	}
> 
> 	if r.atRandomizedElectionTimeout() {
> ```

@arulajmani @nvanbenschoten I just realized that this changes one current behaviour:

Previously, if we are not promotable, we would check again at every tick after the RandomizedElectionTimeout. However, with this change, we will check again on the next RandomizedElectionTimeout.

I could check if we are not promotable, and subtract one from the elapsedElection, but I just want to see what you think and whether you think I should do that or not

File: pkg/raft/raft.go [hunk 1]
```diff
@@ -398,7 +398,7 @@ type raft struct {
 	// the current leader or if the leader is fortified when ticked.
 	//
 	// Invariant: electionElapsed = 0 when r.leadEpoch != 0 on a follower.
-	electionElapsed int
+	electionElapsed uint64
```
Comment (nvanbenschoten) [inline pkg/raft/raft.go:401]: We're performing arithmetic on these fields, so I think we should make them `int64`s, not `uint64`s. Unsigned integers are generally frowned upon as a means of encoding "non-negative semantics". See https://graphitemaster.github.io/aau/ and https://www.learncpp.com/cpp-tutorial/unsigned-integers-and-why-to-avoid-them for some discussion why. I recall there was also a similar best practice at Google. They don't actually enforce that values don't go negative, they just underflow and cause such bugs to be harder to catch. It's also common to want to use a negative offset when manipulating values that should not go negative.

File: pkg/base/config.go [hunk 2]
```diff
@@ -524,15 +524,15 @@ type RaftConfig struct {
 	// an election. The actual election timeout is randomized by each replica to
 	// between 1-2 election timeouts. This value is inherited by individual stores
 	// unless overridden.
-	RaftElectionTimeoutTicks int
+	RaftElectionTimeoutTicks uint64
 
 	// RaftReproposalTimeoutTicks is the number of ticks before reproposing a Raft
 	// command. This also specifies the number of ticks between each reproposal
 	// check, so the actual timeout is 1-2 times this value.
 	RaftReproposalTimeoutTicks int
```
Comment (nvanbenschoten) [inline pkg/base/config.go:532]: If we're going to change the data type for "tick" units, let's do so everywhere.

File: pkg/base/config.go [hunk 3]
```diff
@@ -633,10 +633,10 @@ func (cfg *RaftConfig) SetDefaults() {
 		cfg.RaftTickInterval = defaultRaftTickInterval
 	}
 	if cfg.RaftElectionTimeoutTicks == 0 {
-		cfg.RaftElectionTimeoutTicks = defaultRaftElectionTimeoutTicks
+		cfg.RaftElectionTimeoutTicks = uint64(defaultRaftElectionTimeoutTicks)
```
Comment (nvanbenschoten) [inline pkg/base/config.go:636]: Can we change the type of `defaultRaftElectionTimeoutTicks` and `defaultRaftHeartbeatIntervalTicks`? `EnvOrDefaultInt64` exists, if you agree with my other comment.

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/base/config.go` line 532 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133418#gh-1816955050:-OA4YU2X-TiZrgSJZfOe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653dc82c9d4b84582ad5241a9b0eff31638320e/pkg/base/config.go#L532)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we're going to change the data type for "tick" units, let's do so everywhere.
</blockquote></details>

Done.
___
*[`pkg/base/config.go` line 636 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133418#gh-1816956336:-OA4YV0dDHjUxrtkgS-D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653dc82c9d4b84582ad5241a9b0eff31638320e/pkg/base/config.go#L636)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we change the type of `defaultRaftElectionTimeoutTicks` and `defaultRaftHeartbeatIntervalTicks`? `EnvOrDefaultInt64` exists, if you agree with my other comment.
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 401 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133418#gh-1816952127:-OA4D-u0B50ykZpoGrDN:b-epa4a9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653dc82c9d4b84582ad5241a9b0eff31638320e/pkg/raft/raft.go#L401)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We're performing arithmetic on these fields, so I think we should make them `int64`s, not `uint64`s. Unsigned integers are generally frowned upon as a means of encoding "non-negative semantics". See https://graphitemaster.github.io/aau/ and https://www.learncpp.com/cpp-tutorial/unsigned-integers-and-why-to-avoid-them for some discussion why. I recall there was also a similar best practice at Google. They don't actually enforce that values don't go negative, they just underflow and cause such bugs to be harder to catch. It's also common to want to use a negative offset when manipulating values that should not go negative.
</blockquote></details>

Yeah I had the same concern yesterday after I pushed this PR. I will change it to int64 to be on the safer side

**Review by nvb — COMMENTED** [review]
> Reviewed 19 of 19 files at r4, 3 of 3 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @iskettaneh)

___
*[`-- commits` line 7 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJDY5V2y89aNDfW8v9):*
"int64"
___
*[`-- commits` line 28 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJHdO7BuuKYVjdth06):*
"introduction of"
___
*[`pkg/raft/raft.go` line 83 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJHAcu9coCb0tH5RyG:-OAJHAcu9coCb0tH5RyH:bpofp1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f7b3999ae7fa5fc02cb70a99d4e759a833c8b1/pkg/raft/raft.go#L83)):*
> ```Go
> }
> 
> func (r *lockedRand) Intn64(n int64) int64 {
> ```

The `math/rand` package calls this `Int63n` (https://pkg.go.dev/math/rand#Rand.Int63n), so we should do the same here.
___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJIm0d-JPTH0xGidYK:bx1roza) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
> ```Go
> 	uncommittedSize entryPayloadSize
> 
> 	// electionElapsed is the number of ticks since we last reached the
> ```

We will need to update this comment.

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 83 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJHAcu9coCb0tH5RyG:-OAJMAsX6g6MOKjbwNU1:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/73f7b3999ae7fa5fc02cb70a99d4e759a833c8b1/pkg/raft/raft.go#L83)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The `math/rand` package calls this `Int63n` (https://pkg.go.dev/math/rand#Rand.Int63n), so we should do the same here.
</blockquote></details>

Done!

**Review by nvb — COMMENTED** [review]
> Reviewed 2 of 3 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJRuhp9EMs2rMDQgj_:b-phtd5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We will need to update this comment.
</blockquote></details>

"or if the leader is fortified when ticked"

I've read this part (which was there before) a few times and don't understand it. Do you?

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJSNUHD03crZDzdwu6:b5epgdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"or if the leader is fortified when ticked"

I've read this part (which was there before) a few times and don't understand it. Do you?
</blockquote></details>

My understanding is this: for a follower, as long as it's fortifying a leader, the elapsedElection is zero.
Basically, elapsedElection is set to zero when are fortifying a leader, and we don't increment the elapsedElection if we are fortifying a leader.

I think a simpler sentence would be: "elapsedElection is set to 0 for followers who are fortifying a leader". WDYT?

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJbPO-1GzU9P5qbEtm:boodzdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
> for a follower, as long as it's fortifying a leader, the elapsedElection is zero.

Agreed, I just don't understand the sentence. Maybe I'm reading it wrong. What you suggest sounds good (and is captured in the Invariant mentioned below), although it might be nice to not have two sentences which directly contradict each other back-to-back. 

How about "For followers, it is the number of ticks since they last received a valid message from the from the current leader, unless the follower is fortifying a leader (`leadEpoch != 0`), in which case it is always set to 0."

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 395 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-OAJIm0d-JPTH0xGidYJ:-OAJtS7-5o3UQ0tQ4ftP:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/90dac76795db8f32b1d983b6c45446f407ffdc7b/pkg/raft/raft.go#L395)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> for a follower, as long as it's fortifying a leader, the elapsedElection is zero.

Agreed, I just don't understand the sentence. Maybe I'm reading it wrong. What you suggest sounds good (and is captured in the Invariant mentioned below), although it might be nice to not have two sentences which directly contradict each other back-to-back. 

How about "For followers, it is the number of ticks since they last received a valid message from the from the current leader, unless the follower is fortifying a leader (`leadEpoch != 0`), in which case it is always set to 0."
</blockquote></details>

Done!

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133418#-:-OAK5X6ZE2p9sH0SejUl:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

**Comment by iskettaneh** [comment]
TFTR!

bors r+

File: pkg/raft/raft.go [hunk 4]
```diff
@@ -1184,7 +1184,7 @@ func (r *raft) tickElection() {
 		r.electionElapsed++
 	}
 
-	if r.promotable() && r.pastElectionTimeout() {
```
Comment (pav-kv) [inline pkg/raft/raft.go:1187]: From you second commit description: 
> If a replica is not promotable and we are past the randomizedElectionTimeout, we will keep try to campaign on every tick. 

I’m not following why we do that. The `promotable()` check ensures 3 things:
1. we are part of the active configuration. If not we don’t have a chance to win - nodoby will vote for us
2. we are not a learner. If we are, we aren’t a voter and shouldn’t campaign
3. we aren’t processing an incoming snapshot. I’m less clear on why we can’t campaign if we are

At least in cases 1 and 2, why would we proceed to campaign here?

Are we relying on the fact that `hup` method will check `promotable()` down the stack and not campaign anyway?

File: pkg/raft/raft.go [hunk 5]
```diff
@@ -2647,11 +2647,11 @@ func (r *raft) loadState(state pb.HardState) {
...
 
-// pastElectionTimeout returns true if r.electionElapsed is greater
-// than or equal to the randomized election timeout in
-// [electiontimeout, 2 * electiontimeout - 1].
-func (r *raft) pastElectionTimeout() bool {
-	return r.electionElapsed >= r.randomizedElectionTimeout
+// atRandomizedElectionTimeout returns true if r.electionElapsed modulo the
+// r.randomizedElectionTimeout is equal to 0. This means that at every
+// r.randomizedElectionTimeout period, this method will return true once.
+func (r *raft) atRandomizedElectionTimeout() bool {
```
Comment (pav-kv) [inline pkg/raft/raft.go:2653]: The caller of this doesn't check `promotable()` anymore, so what happens in the following case?

1. tick `randomizedElectionTimeout`: we call `hup`, but `promotable()` is false, so we don't campaign
2. tick `randomizedElectionTimeout+1`: we became promotable, but we will have to wait out another `randomizedElectionTimeout` to campaign

Is this intended?
